sqldelight-runtime/src/main/java/com/squareup/sqldelight/SqlDelightDatabase.kt

```diff
@@ -1,93 +1,81 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package com.squareup.sqldelight
 
-import android.arch.persistence.db.SupportSQLiteOpenHelper
-import android.database.sqlite.SQLiteTransactionListener
-import java.io.Closeable
-import java.util.LinkedHashSet
-import java.util.concurrent.TimeUnit
+import com.squareup.sqldelight.SqlDelightDatabase.Transaction
+import com.squareup.sqldelight.db.SqlDatabaseHelper
 
 /**
  * A transaction-aware [SupportSQLiteOpenHelper] wrapper which can queue work to be performed after
  * the active [Transaction] ends with [afterTransaction].
  */
 abstract class SqlDelightDatabase(
-    internal val helper: SupportSQLiteOpenHelper
-) : Transactor, Closeable {
-
-  private val transactions = ThreadLocal<SqliteTransaction>()
-
-  private val transaction = object : Transaction {
-    override fun markSuccessful() {
-      helper.writableDatabase.setTransactionSuccessful()
-    }
-
-    override fun yieldIfContendedSafely(): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely()
-    }
-
-    override fun yieldIfContendedSafely(sleepAmount: Long, sleepUnit: TimeUnit): Boolean {
-      return helper.writableDatabase.yieldIfContendedSafely(sleepUnit.toMillis(sleepAmount))
-    }
-
-    override fun end() {
-      val transaction = transactions.get() ?: throw IllegalStateException("Not in transaction.")
-      transactions.set(transaction.parent)
-      helper.writableDatabase.endTransaction()
-      if (transaction.commit) {
-        transaction.forEach { function -> afterTransaction(function) }
-      }
-    }
-
-    override fun close() {
-      end()
-    }
-  }
-
-  override fun newTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListener(transaction)
-
-    return this.transaction
-  }
-
-  override fun newNonExclusiveTransaction(): Transaction {
-    val transaction = SqliteTransaction(transactions.get())
-    transactions.set(transaction)
-    helper.writableDatabase.beginTransactionWithListenerNonExclusive(transaction)
+  internal val helper: SqlDatabaseHelper
+) {
+  private val transactions = ThreadLocal<Transaction>()
+  private val actions = ArrayList<() -> Unit>()
 
-    return this.transaction
-  }
-
-  override fun afterTransaction(function: () -> Unit) {
+  fun afterTransaction(function: () -> Unit) {
```

Comment:
afterCommit ?

Reply:
... on the server we have afterRollback also

Reply:
added `afterRollback` and some tests for it.